File: Web3Wallet/Domain/Entities/Network.swift
//
//  Network.swift
//  Web3Wallet
//
//  Created by Zy on 2025/10/27.
//  Copyright © 2025 Web3Wallet. All rights reserved.
//

import Foundation

/// Network configuration entity
/// Define supported blockchain networks and their configuration information
struct Network: Equatable, Codable {
    let id: String
    let name: String
    let chainId: Int
    let rpcURL: String
    let explorerBase: String
    let nativeCurrency: Currency
    let isTestnet: Bool
    
    static let ethereumMainnet = Network(
        id: "ethereum_mainnet",
        name: "Ethereum Mainnet",
        chainId: 1,
        rpcURL: "https://mainnet.infura.io/v3/1b7ed1f23d854cd99b816a1b6ea27b12",
        explorerBase: "https://etherscan.io",
        nativeCurrency: Currency(
            symbol: "ETH",
            name: "Ethereum",
            decimals: 18,
            contractAddress: nil
        ),
        isTestnet: false
    )
    
    static let sepolia = Network(
        id: "sepolia",
        name: "Sepolia Testnet",
        chainId: 11155111,
        rpcURL: "https://sepolia.infura.io/v3/1b7ed1f23d854cd99b816a1b6ea27b12",
        explorerBase: "https://sepolia.etherscan.io",
        nativeCurrency: Currency(
            symbol: "ETH",
            name: "Ethereum",
            decimals: 18,
            contractAddress: nil
        ),
        isTestnet: true
    )
    
    static let supportedNetworks: [Network] = [.ethereumMainnet, .sepolia]
}

File: ReactNativeWallet/src/domain/entities/Network.ts
import { CURRENCIES } from './Currency';

export type Network = {
  id: string;
  name: string;
  chainId: number;
  rpcUrl: string;
  explorerBaseUrl: string;
  nativeCurrency: typeof CURRENCIES.eth;
  isTestnet: boolean;
};

const ETHEREUM_MAINNET: Network = {
  id: 'ethereum_mainnet',
  name: 'Ethereum Mainnet',
  chainId: 1,
  rpcUrl: 'https://mainnet.infura.io/v3/1b7ed1f23d854cd99b816a1b6ea27b12',
  explorerBaseUrl: 'https://etherscan.io',
  nativeCurrency: CURRENCIES.eth,
  isTestnet: false,
};

const SEPOLIA_TESTNET: Network = {
  id: 'sepolia',
  name: 'Sepolia Testnet',
  chainId: 11155111,
  rpcUrl: 'https://sepolia.infura.io/v3/1b7ed1f23d854cd99b816a1b6ea27b12',
  explorerBaseUrl: 'https://sepolia.etherscan.io',
  nativeCurrency: CURRENCIES.eth,
  isTestnet: true,
};

export const NETWORKS = {
  ethereumMainnet: ETHEREUM_MAINNET,
  sepolia: SEPOLIA_TESTNET,
} as const;

export const SUPPORTED_NETWORKS: Network[] = [
  NETWORKS.ethereumMainnet,
  NETWORKS.sepolia,
];
